core/base/batch_multi_vector.cpp

```diff
@@ -0,0 +1,256 @@
+    for (size_type b = 0; b < alpha->get_num_batch_entries(); ++b) {
+        if (alpha->get_common_size()[1] != 1) {
+            // different alpha for each column
+            GKO_ASSERT_EQUAL_COLS(this->get_common_size(),
+                                  alpha->get_common_size());
+        }
+    }
+    GKO_ASSERT_EQ(b->get_num_batch_entries(), this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(this->get_common_size(), b->get_common_size());
+
+    this->get_executor()->run(
+        batch_multi_vector::make_add_scaled(alpha, b, this));
+}
+
+
+inline const batch_dim<2> get_col_sizes(const batch_dim<2>& sizes)
+{
+    return batch_dim<2>(sizes.get_num_batch_entries(),
+                        dim<2>(1, sizes.get_common_size()[1]));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::compute_dot_impl(
+    const BatchMultiVector<ValueType>* b,
+    BatchMultiVector<ValueType>* result) const
+{
+    GKO_ASSERT_EQ(b->get_num_batch_entries(), this->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(this->get_common_size(), b->get_common_size());
+    GKO_ASSERT_EQ(this->get_num_batch_entries(),
+                  result->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(
+        result->get_common_size(),
+        get_col_sizes(this->get_size()).get_common_size());
+    this->get_executor()->run(
+        batch_multi_vector::make_compute_dot(this, b, result));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::compute_norm2_impl(
+    BatchMultiVector<remove_complex<ValueType>>* result) const
+{
+    GKO_ASSERT_EQ(this->get_num_batch_entries(),
+                  result->get_num_batch_entries());
+    GKO_ASSERT_EQUAL_DIMENSIONS(
+        result->get_common_size(),
+        get_col_sizes(this->get_size()).get_common_size());
+    this->get_executor()->run(batch_multi_vector::make_compute_norm2(
+        as<BatchMultiVector<ValueType>>(this), result));
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::convert_to(
+    BatchMultiVector<next_precision<ValueType>>* result) const
+{
+    result->values_ = this->values_;
+    result->set_size(this->get_size());
+}
+
+
+template <typename ValueType>
+void BatchMultiVector<ValueType>::move_to(
+    BatchMultiVector<next_precision<ValueType>>* result)
+{
+    this->convert_to(result);
+}
+
+
+template <typename MatrixType, typename MatrixData>
+inline void read_impl(MatrixType* mtx, const std::vector<MatrixData>& data)
+{
+    auto common_size = data[0].size;
+    auto batch_size = batch_dim<2>(data.size(), common_size);
+    size_type ind = 0;
+    for (const auto& b : data) {
+        auto b_size = b.size;
+        GKO_ASSERT_EQUAL_DIMENSIONS(common_size, b_size);
+    }
+    auto tmp =
+        MatrixType::create(mtx->get_executor()->get_master(), batch_size);
+    for (size_type b = 0; b < data.size(); ++b) {
+        size_type ind = 0;
+        for (size_type row = 0; row < data[b].size[0]; ++row) {
```

Comment:
If you fill the array with zeros, you wouldn't need this nested for-loop + if branch, which would make this clearer

Reply:
Since you marked this as resolved, will you change it, or did I miss something?

Reply:
Are you checking the latest change ? This file has changed a lot, so line numbers dont match anymore

Reply:
the matrix data does not contain m * n entries.
It currently goes over all m * n and fill the value into the matrices. The ind will be m*n in the end.
Because you already initial the vectors to zero, you only need to go over the mat_data.size to fill the value into the matrix.